core/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -100,16 +99,18 @@ public NodePath(Path path, Environment environment) throws IOException {
 
         /**
          * Resolves the given shards directory against this NodePath
+         * ${data.paths}/nodes/{node.id}/indices/{index.uuid}/{shard.id}
          */
         public Path resolve(ShardId shardId) {
-            return resolve(shardId.getIndex()).resolve(Integer.toString(shardId.id()));
+            return resolve(shardId.getIndex().getUUID()).resolve(Integer.toString(shardId.id()));
```

Comment:
I think we should add a `Index#getPathIdentifier()` to have a single place to calculate it?